File: implementations/ordered-map/rust/src/ordered_map.rs
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ClassToken {
    A,
    B,
    C,
}

impl ClassToken {
    pub fn from_key(key: &str) -> Result<Self, String> {
        match key {
            "A" | "a" => Ok(Self::A),
            "B" | "b" => Ok(Self::B),
            "C" | "c" => Ok(Self::C),
            _ => Err("key is outside the selected profile domain".to_owned()),
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::A => "a",
            Self::B => "b",
            Self::C => "c",
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct OrderedMap {
    entries: Vec<(ClassToken, i64)>,
}

impl OrderedMap {
    pub fn empty() -> Self {
        Self { entries: Vec::new() }
    }

    pub fn put(&self, key: &str, value: i64) -> Result<Self, String> {
        let token = ClassToken::from_key(key)?;
        let mut entries = self.entries.clone();
        if let Some(position) = entries.iter().position(|(item, _)| *item == token) {
            entries[position].1 = value;
        } else {
            entries.push((token, value));
        }
        Ok(Self { entries })
    }

    pub fn lookup(&self, key: &str) -> Result<Option<i64>, String> {
        let token = ClassToken::from_key(key)?;
        Ok(self
            .entries
            .iter()
            .find(|(item, _)| *item == token)
            .map(|(_, value)| *value))
    }

    pub fn entries(&self) -> &[(ClassToken, i64)] {
        &self.entries
    }
}

#[cfg(test)]
mod tests {
    use super::OrderedMap;

    #[test]
    fn replacement_retains_position_and_sources() {
        let empty = OrderedMap::empty();
        let one = empty.put("A", 1).unwrap();
        let two = one.put("B", 2).unwrap();
        let replaced = two.put("a", -1).unwrap();

        assert_eq!(None, empty.lookup("a").unwrap());
        assert_eq!(Some(1), one.lookup("A").unwrap());
        assert_eq!([("a", -1), ("b", 2)], visible(&replaced)[..]);
        assert_eq!([("a", 1), ("b", 2)], visible(&two)[..]);
    }

    fn visible(map: &OrderedMap) -> Vec<(&'static str, i64)> {
        map.entries()
            .iter()
            .map(|(token, value)| (token.as_str(), *value))
            .collect()
    }
}

File: implementations/ordered-map/typescript/ordered_map.ts
export type ClassToken = "a" | "b" | "c";

export interface Entry {
  readonly class: ClassToken;
  readonly value: number;
}

export class OrderedMap {
  private readonly values: ReadonlyMap<ClassToken, number>;
  private readonly order: readonly ClassToken[];

  private constructor(
    values: ReadonlyMap<ClassToken, number>,
    order: readonly ClassToken[],
  ) {
    this.values = values;
    this.order = order;
    Object.freeze(this);
  }

  static empty(): OrderedMap {
    return new OrderedMap(new Map(), Object.freeze([]));
  }

  put(key: string, value: number): OrderedMap {
    const token = classify(key);
    const values = new Map(this.values);
    values.set(token, value);
    const order = this.values.has(token)
      ? this.order
      : Object.freeze([...this.order, token]);
    return new OrderedMap(values, order);
  }

  lookup(key: string): number | undefined {
    return this.values.get(classify(key));
  }

  entries(): readonly Entry[] {
    return Object.freeze(this.order.map((token) =>
      Object.freeze({ class: token, value: this.values.get(token)! })
    ));
  }
}

function classify(key: string): ClassToken {
  const token = key.toLowerCase();
  if (token !== "a" && token !== "b" && token !== "c") {
    throw new Error("key is outside the selected profile domain");
  }
  return token;
}
